File: searching/AdvancedSearch.java
/**
 * Advanced Search Algorithms - Java Implementation
 * ================================================
 *
 * Comprehensive collection of advanced search algorithms including:
 * 1. Jump Search
 * 2. Fibonacci Search
 * 3. Interpolation Search
 * 4. Exponential Search
 * 5. Parallel Search
 * 6. Geometric Search (KD-Tree)
 * 7. Fuzzy Search (Levenshtein)
 *
 * Features:
 * - Generic implementations with Comparator support
 * - Cache-friendly algorithms
 * - Parallel implementations using Fork/Join
 * - Performance-optimized for JVM
 *
 * Compilation: javac AdvancedSearch.java
 * Usage: java AdvancedSearch
 */

import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.stream.*;

public class AdvancedSearch {

    // =========================================================================
    // 1. JUMP SEARCH
    // =========================================================================

    /**
     * Jump Search - Block-based search algorithm
     *
     * Time: O(√n), Space: O(1)
     *
     * WHEN TO USE:
     * - Comparisons are expensive (complex objects)
     * - Sequential access patterns preferred
     * - Want fewer comparisons than binary search
     *
     * @param arr Sorted array
     * @param target Element to find
     * @return Index of target or -1
     */
    public static <T extends Comparable<T>> int jumpSearch(T[] arr, T target) {
        if (arr == null || arr.length == 0) return -1;

        int n = arr.length;
        int step = (int) Math.sqrt(n);  // Optimal step size
        int prev = 0;

        // Jump ahead to find block containing target
        while (prev < n && arr[Math.min(step, n) - 1].compareTo(target) < 0) {
            prev = step;
            step += (int) Math.sqrt(n);

            if (prev >= n) return -1;
        }

        // Linear search in the block
        while (prev < n && arr[prev].compareTo(target) < 0) {
            prev++;
            if (prev == Math.min(step, n)) return -1;
        }

        // Check if element found
        if (prev < n && arr[prev].compareTo(target) == 0) {
            return prev;
        }

        return -1;
    }

    // =========================================================================
    // 2. FIBONACCI SEARCH
    // =========================================================================

    /**
     * Fibonacci Search - Using Fibonacci numbers for division
     *
     * Time: O(log n), Space: O(1)
     *
     * ADVANTAGES:
     * - No division operations (only addition/subtraction)
     * - Good for systems where division is expensive
     * - Cache-friendly due to smaller jumps
     *
     * @param arr Sorted array
     * @param target Element to find
     * @return Index of target or -1
     */
    public static <T extends Comparable<T>> int fibonacciSearch(T[] arr, T target) {
        if (arr == null || arr.length == 0) return -1;

        int n = arr.length;

        // Initialize Fibonacci numbers
        int fibM2 = 0;  // (m-2)th Fibonacci number
        int fibM1 = 1;  // (m-1)th Fibonacci number
        int fibM = fibM2 + fibM1;  // mth Fibonacci number

        // Find smallest Fibonacci >= n
        while (fibM < n) {
            fibM2 = fibM1;
            fibM1 = fibM;
            fibM = fibM2 + fibM1;
        }

        int offset = -1;

        while (fibM > 1) {
            // Check if fibM2 is valid index
            int i = Math.min(offset + fibM2, n - 1);

            int cmp = arr[i].compareTo(target);

            if (cmp < 0) {
                // Move offset forward
                fibM = fibM1;
                fibM1 = fibM2;
                fibM2 = fibM - fibM1;
                offset = i;
            } else if (cmp > 0) {
                // Move Fibonacci numbers down
                fibM = fibM2;
                fibM1 = fibM1 - fibM2;
                fibM2 = fibM - fibM1;
            } else {
                return i;
            }
        }

        // Check last element
        if (fibM1 == 1 && offset + 1 < n && arr[offset + 1].compareTo(target) == 0) {
            return offset + 1;
        }

        return -1;
    }

    // =========================================================================
    // 3. INTERPOLATION SEARCH
    // =========================================================================

    /**
     * Interpolation Search - Value-based position estimation
     *
     * Time: O(log log n) average for uniform data, O(n) worst
     * Space: O(1)
     *
     * WHEN TO USE:
     * - Uniformly distributed numerical data
     * - Large sorted arrays
     * - Known data distribution
     *
     * @param arr Sorted integer array
     * @param target Element to find
     * @return Index of target or -1
     */
    public static int interpolationSearch(int[] arr, int target) {
        if (arr == null || arr.length == 0) return -1;

        int left = 0;
        int right = arr.length - 1;

        while (left <= right && target >= arr[left] && target <= arr[right]) {
            if (left == right) {
                return arr[left] == target ? left : -1;
            }

            // Interpolation formula
            int pos = left + ((target - arr[left]) * (right - left)) /
                             (arr[right] - arr[left]);

            if (arr[pos] == target) {
                return pos;
            } else if (arr[pos] < target) {
                left = pos + 1;
            } else {
                right = pos - 1;
            }
        }

        return -1;
    }

    // =========================================================================
    // 4. EXPONENTIAL SEARCH
    // =========================================================================

    /**
     * Exponential Search - Finds range then binary searches
     *
     * Time: O(log n), Space: O(1)
     *
     * WHEN TO USE:
     * - Unbounded/infinite arrays
     * - Target expected near beginning
     * - Unknown array size
     *
     * @param arr Sorted array
     * @param target Element to find
     * @return Index of target or -1
     */
    public static <T extends Comparable<T>> int exponentialSearch(T[] arr, T target) {
        if (arr == null || arr.length == 0) return -1;
        if (arr[0].compareTo(target) == 0) return 0;

        // Find range for binary search
        int i = 1;
        while (i < arr.length && arr[i].compareTo(target) <= 0) {
            i *= 2;
        }

        // Binary search in range [i/2, min(i, n-1)]
        return binarySearch(arr, target, i / 2, Math.min(i, arr.length - 1));
    }

    private static <T extends Comparable<T>> int binarySearch(
            T[] arr, T target, int left, int right) {
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int cmp = arr[mid].compareTo(target);

            if (cmp == 0) return mid;
            else if (cmp < 0) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }

    // =========================================================================
    // 5. PARALLEL SEARCH
    // =========================================================================

    /**
     * Parallel Search using Fork/Join framework
     *
     * Time: O(n / cores) with parallelism
     * Space: O(cores)
     *
     * WHEN TO USE:
     * - Very large arrays (millions of elements)
     * - Multi-core CPUs available
     * - Can work on unsorted data
     *
     * @param arr Array to search (can be unsorted)
     * @param target Element to find
     * @return Index of target or -1
     */
    public static <T> int parallelSearch(T[] arr, T target) {
        if (arr == null || arr.length == 0) return -1;

        ForkJoinPool pool = ForkJoinPool.commonPool();
        return pool.invoke(new ParallelSearchTask<>(arr, target, 0, arr.length));
    }

    static class ParallelSearchTask<T> extends RecursiveTask<Integer> {
        private static final int THRESHOLD = 1000;
        private final T[] arr;
        private final T target;
        private final int start;
        private final int end;

        ParallelSearchTask(T[] arr, T target, int start, int end) {
            this.arr = arr;
            this.target = target;
            this.start = start;
            this.end = end;
        }

        @Override
        protected Integer compute() {
            int length = end - start;

            if (length <= THRESHOLD) {
                // Sequential search for small ranges
                for (int i = start; i < end; i++) {
                    if (Objects.equals(arr[i], target)) {
                        return i;
                    }
                }
                return -1;
            }

            // Split task
            int mid = start + length / 2;
            ParallelSearchTask<T> leftTask = new ParallelSearchTask<>(arr, target, start, mid);
            ParallelSearchTask<T> rightTask = new ParallelSearchTask<>(arr, target, mid, end);

            leftTask.fork();
            int rightResult = rightTask.compute();
            int leftResult = leftTask.join();

            // Return first found result
            if (leftResult != -1) return leftResult;
            return rightResult;
        }
    }

    // =========================================================================
    // 6. KD-TREE FOR GEOMETRIC SEARCH
    // =========================================================================

    /**
     * KD-Tree for multidimensional search
     *
     * Time:
     * - Build: O(n log n)
     * - Search: O(log n) average
     * - Nearest neighbor: O(log n) average
     *
     * WHEN TO USE:
     * - Nearest neighbor in multiple dimensions
     * - Range queries in 2D/3D space
     * - GIS applications
     * - Computer graphics
     */
    static class KDTree {
        private static class Node {
            double[] point;
            Node left, right;
            int axis;

            Node(double[] point, int axis) {
                this.point = point;
                this.axis = axis;
            }
        }

        private Node root;
        private int k;  // Number of dimensions

        public KDTree(double[][] points) {
            if (points.length == 0) return;
            this.k = points[0].length;
            this.root = build(points, 0);
        }

        private Node build(double[][] points, int depth) {
            if (points.length == 0) return null;

            int axis = depth % k;

            // Sort by current axis
            Arrays.sort(points, Comparator.comparingDouble(p -> p[axis]));
            int median = points.length / 2;

            Node node = new Node(points[median], axis);
            node.left = build(Arrays.copyOfRange(points, 0, median), depth + 1);
            node.right = build(Arrays.copyOfRange(points, median + 1, points.length), depth + 1);

            return node;
        }

        /**
         * Find nearest neighbor to query point
         */
        public NearestResult nearestNeighbor(double[] query) {
            if (root == null) return null;

            NearestResult best = new NearestResult(null, Double.MAX_VALUE);
            nearestSearch(root, query, best);
            return best;
        }

        private void nearestSearch(Node node, double[] query, NearestResult best) {
            if (node == null) return;

            double dist = distance(query, node.point);
            if (dist < best.distance) {
                best.point = node.point;
                best.distance = dist;
            }

            double diff = query[node.axis] - node.point[node.axis];
            Node near = diff < 0 ? node.left : node.right;
            Node far = diff < 0 ? node.right : node.left;

            nearestSearch(near, query, best);

            if (Math.abs(diff) < best.distance) {
                nearestSearch(far, query, best);
            }
        }

        private double distance(double[] p1, double[] p2) {
            double sum = 0;
            for (int i = 0; i < k; i++) {
                double diff = p1[i] - p2[i];
                sum += diff * diff;
            }
            return Math.sqrt(sum);
        }

        static class NearestResult {
            double[] point;
            double distance;

            NearestResult(double[] point, double distance) {
                this.point = point;
                this.distance = distance;
            }
        }
    }

    // =========================================================================
    // 7. FUZZY SEARCH (LEVENSHTEIN DISTANCE)
    // =========================================================================

    /**
     * Calculate Levenshtein (edit) distance between two strings
     *
     * Time: O(m * n), Space: O(min(m, n))
     */
    public static int levenshteinDistance(String s1, String s2) {
        if (s1.length() < s2.length()) {
            String temp = s1;
            s1 = s2;
            s2 = temp;
        }

        int m = s1.length();
        int n = s2.length();

        if (n == 0) return m;

        // Use two rows for space optimization
        int[] prevRow = new int[n + 1];
        int[] currRow = new int[n + 1];

        // Initialize first row
        for (int j = 0; j <= n; j++) {
            prevRow[j] = j;
        }

        for (int i = 1; i <= m; i++) {
            currRow[0] = i;

            for (int j = 1; j <= n; j++) {
                int cost = s1.charAt(i - 1) == s2.charAt(j - 1) ? 0 : 1;

                currRow[j] = Math.min(
                    Math.min(prevRow[j] + 1,      // Deletion
                            currRow[j - 1] + 1),  // Insertion
                    prevRow[j - 1] + cost         // Substitution
                );
            }

            // Swap rows
            int[] temp = prevRow;
            prevRow = currRow;
            currRow = temp;
        }

        return prevRow[n];
    }

    /**
     * Fuzzy search - Find approximate matches of pattern in text
     *
     * @param text Text to search in
     * @param pattern Pattern to find
     * @param maxDistance Maximum edit distance allowed
     * @return List of (start, end) positions
     */
    public static List<int[]> fuzzySearch(String text, String pattern, int maxDistance) {
        List<int[]> matches = new ArrayList<>();
        int textLen = text.length();
        int patternLen = pattern.length();

        for (int i = 0; i < textLen - patternLen + maxDistance + 1; i++) {
            for (int windowSize = Math.max(1, patternLen - maxDistance);
                 windowSize <= Math.min(textLen - i, patternLen + maxDistance);
                 windowSize++) {

                String window = text.substring(i, i + windowSize);
                int distance = levenshteinDistance(pattern, window);

                if (distance <= maxDistance) {
                    matches.add(new int[]{i, i + windowSize});
                    break;
                }
            }
        }

        return matches;
    }

    // =========================================================================
    // EXAMPLES AND TESTING
    // =========================================================================

    private static void printSeparator() {
        System.out.println("=".repeat(70));
    }

    private static void exampleJumpSearch() {
        printSeparator();
        System.out.println("EXAMPLE 1: Jump Search");
        printSeparator();

        Integer[] arr = {1, 3, 5, 7, 9, 11, 13, 15, 17, 19};
        System.out.println("Array: " + Arrays.toString(arr));
        System.out.println("Jump search for 11: index " + jumpSearch(arr, 11));
        System.out.println("Jump search for 20: index " + jumpSearch(arr, 20));
        System.out.println();
    }

    private static void exampleFibonacciSearch() {
        printSeparator();
        System.out.println("EXAMPLE 2: Fibonacci Search");
        printSeparator();

        Integer[] arr = {2, 3, 4, 10, 40, 50, 60, 70, 80, 90};
        System.out.println("Array: " + Arrays.toString(arr));
        System.out.println("Fibonacci search for 40: index " + fibonacciSearch(arr, 40));
        System.out.println("Fibonacci search for 100: index " + fibonacciSearch(arr, 100));
        System.out.println();
    }

    private static void exampleKDTree() {
        printSeparator();
        System.out.println("EXAMPLE 3: KD-Tree Geometric Search");
        printSeparator();

        double[][] points = {
            {2, 3}, {5, 4}, {9, 6}, {4, 7}, {8, 1}, {7, 2}
        };

        KDTree tree = new KDTree(points);
        double[] query = {6, 3};

        KDTree.NearestResult result = tree.nearestNeighbor(query);
        System.out.println("Query point: " + Arrays.toString(query));
        System.out.println("Nearest neighbor: " + Arrays.toString(result.point));
        System.out.printf("Distance: %.2f%n", result.distance);
        System.out.println();
    }

    private static void exampleFuzzySearch() {
        printSeparator();
        System.out.println("EXAMPLE 4: Fuzzy String Search");
        printSeparator();

        String text = "The quick brown fox jumps over the lazy dog";
        String pattern = "quik";

        System.out.println("Text: '" + text + "'");
        System.out.println("Pattern: '" + pattern + "' (max distance = 1)");

        List<int[]> matches = fuzzySearch(text, pattern, 1);
        System.out.println("Matches:");
        for (int[] match : matches) {
            System.out.printf("  [%d, %d]: '%s'%n",
                match[0], match[1], text.substring(match[0], match[1]));
        }
        System.out.println();
    }

    private static void exampleParallelSearch() {
        printSeparator();
        System.out.println("EXAMPLE 5: Parallel Search");
        printSeparator();

        // Create large array
        Integer[] arr = new Integer[10000];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = i * 2;
        }

        int target = 5000;
        long start = System.nanoTime();
        int result = parallelSearch(arr, target);
        long time = (System.nanoTime() - start) / 1000;

        System.out.println("Array size: " + arr.length);
        System.out.println("Target: " + target);
        System.out.println("Found at index: " + result);
        System.out.printf("Time: %d microseconds%n", time);
        System.out.println();
    }

    public static void main(String[] args) {
        printSeparator();
        System.out.println("ADVANCED SEARCH ALGORITHMS - JAVA");
        printSeparator();
        System.out.println();

        exampleJumpSearch();
        exampleFibonacciSearch();
        exampleKDTree();
        exampleFuzzySearch();
        exampleParallelSearch();

        printSeparator();
        System.out.println("All examples completed!");
        printSeparator();
    }
}

File: searching/advanced_search.f90
! Advanced Search Algorithms in Modern Fortran
!
! Comprehensive collection of search algorithms:
! 1. Binary Search (iterative and recursive)
! 2. Interpolation Search
! 3. Jump Search
! 4. Exponential Search
! 5. Ternary Search
! 6. Fibonacci Search
!
! Author: Algorithms Multiverse

program advanced_search
    implicit none

    integer, parameter :: n = 1000
    integer :: arr(n), target, result
    integer :: i
    real(8) :: start_time, end_time

    print '(A)', repeat('=', 75)
    print '(A)', '            ADVANCED SEARCH ALGORITHMS IN FORTRAN'
    print '(A)', repeat('=', 75)
    print *

    ! Create sorted array
    do i = 1, n
        arr(i) = i * 2  ! Even numbers: 2, 4, 6, ..., 2000
    end do

    ! Run demonstrations
    call demo_binary_search(arr, n)
    call demo_interpolation_search(arr, n)
    call demo_jump_search(arr, n)
    call demo_exponential_search(arr, n)
    call demo_ternary_search()
    call demo_fibonacci_search(arr, n)

    ! Performance comparison
    call performance_comparison(arr, n)

    ! Summary
    print '(A)', repeat('=', 75)
    print '(A)', 'SEARCH ALGORITHM COMPLEXITY SUMMARY'
    print '(A)', repeat('=', 75)
    print '(A)', 'Algorithm           Best Case    Average      Worst Case   Space'
    print '(A)', repeat('-', 75)
    print '(A)', 'Binary Search       O(1)         O(log n)     O(log n)     O(1)'
    print '(A)', 'Interpolation       O(1)         O(log log n) O(n)         O(1)'
    print '(A)', 'Jump Search         O(1)         O(√n)        O(√n)        O(1)'
    print '(A)', 'Exponential         O(1)         O(log n)     O(log n)     O(1)'
    print '(A)', 'Ternary Search      O(1)         O(log₃ n)    O(log₃ n)    O(log n)'
    print '(A)', 'Fibonacci Search    O(1)         O(log n)     O(log n)     O(1)'
    print '(A)', repeat('=', 75)

contains

    ! ===================================================================
    ! BINARY SEARCH
    ! ===================================================================

    subroutine demo_binary_search(arr, n)
        integer, intent(in) :: arr(:), n
        integer :: target, result

        print '(A)', 'Test 1: Binary Search'
        print '(A)', repeat('-', 75)

        target = 500
        print '(A, I0)', 'Searching for: ', target

        ! Iterative version
        result = binary_search_iterative(arr, n, target)
        call print_result(result, target, "Iterative")

        ! Recursive version
        result = binary_search_recursive(arr, 1, n, target)
        call print_result(result, target, "Recursive")

        print *
        print '(A)', 'Complexity: O(log n)'
        print '(A)', 'Best for: Sorted arrays, most common search algorithm'
        print *
    end subroutine demo_binary_search

    function binary_search_iterative(arr, n, target) result(index)
        integer, intent(in) :: arr(:), n, target
        integer :: index, left, right, mid

        left = 1
        right = n

        do while (left <= right)
            mid = left + (right - left) / 2

            if (arr(mid) == target) then
                index = mid
                return
            else if (arr(mid) < target) then
                left = mid + 1
            else
                right = mid - 1
            end if
        end do

        index = -1  ! Not found
    end function binary_search_iterative

    recursive function binary_search_recursive(arr, left, right, target) result(index)
        integer, intent(in) :: arr(:), left, right, target
        integer :: index, mid

        if (right < left) then
            index = -1  ! Not found
            return
        end if

        mid = left + (right - left) / 2

        if (arr(mid) == target) then
            index = mid
        else if (arr(mid) < target) then
            index = binary_search_recursive(arr, mid + 1, right, target)
        else
            index = binary_search_recursive(arr, left, mid - 1, target)
        end if
    end function binary_search_recursive

    ! ===================================================================
    ! INTERPOLATION SEARCH
    ! ===================================================================

    subroutine demo_interpolation_search(arr, n)
        integer, intent(in) :: arr(:), n
        integer :: target, result

        print '(A)', 'Test 2: Interpolation Search'
        print '(A)', repeat('-', 75)

        target = 500
        print '(A, I0)', 'Searching for: ', target

        result = interpolation_search(arr, n, target)
        call print_result(result, target, "Interpolation")

        print *
        print '(A)', 'Complexity: O(log log n) for uniformly distributed data'
        print '(A)', 'Best for: Large sorted arrays with uniform distribution'
        print '(A)', 'Uses interpolation formula to estimate position'
        print *
    end subroutine demo_interpolation_search

    function interpolation_search(arr, n, target) result(index)
        integer, intent(in) :: arr(:), n, target
        integer :: index, low, high, pos

        low = 1
        high = n
        index = -1

        do while (low <= high .and. target >= arr(low) .and. target <= arr(high))
            if (low == high) then
                if (arr(low) == target) then
                    index = low
                end if
                return
            end if

            ! Interpolation formula
            pos = low + ((target - arr(low)) * (high - low)) / (arr(high) - arr(low))

            ! Ensure pos is within bounds
            if (pos < low) pos = low
            if (pos > high) pos = high

            if (arr(pos) == target) then
                index = pos
                return
            else if (arr(pos) < target) then
                low = pos + 1
            else
                high = pos - 1
            end if
        end do
    end function interpolation_search

    ! ===================================================================
    ! JUMP SEARCH
    ! ===================================================================

    subroutine demo_jump_search(arr, n)
        integer, intent(in) :: arr(:), n
        integer :: target, result

        print '(A)', 'Test 3: Jump Search'
        print '(A)', repeat('-', 75)

        target = 500
        print '(A, I0)', 'Searching for: ', target

        result = jump_search(arr, n, target)
        call print_result(result, target, "Jump")

        print *
        print '(A)', 'Complexity: O(√n)'
        print '(A)', 'Best for: Sorted arrays when binary search overhead is high'
        print '(A)', 'Jump size: √n for optimal performance'
        print *
    end subroutine demo_jump_search

    function jump_search(arr, n, target) result(index)
        integer, intent(in) :: arr(:), n, target
        integer :: index, step, prev, i

        step = int(sqrt(real(n)))
        prev = 1
        index = -1

        ! Find block where element may be present
        do while (arr(min(step, n)) < target)
            prev = step
            step = step + int(sqrt(real(n)))
            if (prev >= n) return
        end do

        ! Linear search in block
        do i = prev, min(step, n)
            if (arr(i) == target) then
                index = i
                return
            end if
        end do
    end function jump_search

    ! ===================================================================
    ! EXPONENTIAL SEARCH
    ! ===================================================================

    subroutine demo_exponential_search(arr, n)
        integer, intent(in) :: arr(:), n
        integer :: target, result

        print '(A)', 'Test 4: Exponential Search'
        print '(A)', repeat('-', 75)

        target = 500
        print '(A, I0)', 'Searching for: ', target

        result = exponential_search(arr, n, target)
        call print_result(result, target, "Exponential")

        print *
        print '(A)', 'Complexity: O(log n)'
        print '(A)', 'Best for: Unbounded/infinite arrays, unknown size'
        print '(A)', 'Strategy: Find range, then binary search'
        print *
    end subroutine demo_exponential_search

    function exponential_search(arr, n, target) result(index)
        integer, intent(in) :: arr(:), n, target
        integer :: index, bound

        ! If element is at first position
        if (arr(1) == target) then
            index = 1
            return
        end if

        ! Find range for binary search
        bound = 1
        do while (bound < n .and. arr(bound) < target)
            bound = bound * 2
        end do

        ! Binary search in found range
        index = binary_search_recursive(arr, bound / 2 + 1, min(bound, n), target)
    end function exponential_search

    ! ===================================================================
    ! TERNARY SEARCH (for unimodal functions)
    ! ===================================================================

    subroutine demo_ternary_search()
        real(8) :: left, right, max_val, max_pos

        print '(A)', 'Test 5: Ternary Search (for unimodal functions)'
        print '(A)', repeat('-', 75)

        left = -10.0
        right = 10.0

        max_pos = ternary_search(left, right)
        max_val = test_function(max_pos)

        print '(A, F10.6)', 'Maximum found at x = ', max_pos
        print '(A, F10.6)', 'Maximum value = ', max_val
        print *
        print '(A)', 'Complexity: O(log₃ n)'
        print '(A)', 'Best for: Finding maximum/minimum of unimodal function'
        print '(A)', 'Divides range into 3 parts instead of 2'
        print *
    end subroutine demo_ternary_search

    function ternary_search(left, right) result(max_pos)
        real(8), intent(in) :: left, right
        real(8) :: max_pos
        real(8) :: l, r, mid1, mid2
        real(8), parameter :: epsilon = 1.0e-6

        l = left
        r = right

        do while (r - l > epsilon)
            mid1 = l + (r - l) / 3.0
            mid2 = r - (r - l) / 3.0

            if (test_function(mid1) < test_function(mid2)) then
                l = mid1
            else
                r = mid2
            end if
        end do

        max_pos = (l + r) / 2.0
    end function ternary_search

    ! Test function: parabola with maximum at x=0
    function test_function(x) result(y)
        real(8), intent(in) :: x
        real(8) :: y
        y = -x * x + 10.0
    end function test_function

    ! ===================================================================
    ! FIBONACCI SEARCH
    ! ===================================================================

    subroutine demo_fibonacci_search(arr, n)
        integer, intent(in) :: arr(:), n
        integer :: target, result

        print '(A)', 'Test 6: Fibonacci Search'
        print '(A)', repeat('-', 75)

        target = 500
        print '(A, I0)', 'Searching for: ', target

        result = fibonacci_search(arr, n, target)
        call print_result(result, target, "Fibonacci")

        print *
        print '(A)', 'Complexity: O(log n)'
        print '(A)', 'Best for: Systems where division is expensive'
        print '(A)', 'Uses Fibonacci numbers to divide array'
        print *
    end subroutine demo_fibonacci_search

    function fibonacci_search(arr, n, target) result(index)
        integer, intent(in) :: arr(:), n, target
        integer :: index
        integer :: fib_m2, fib_m1, fib_m, offset, i

        ! Initialize Fibonacci numbers
        fib_m2 = 0
        fib_m1 = 1
        fib_m = fib_m2 + fib_m1

        ! Find smallest Fibonacci >= n
        do while (fib_m < n)
            fib_m2 = fib_m1
            fib_m1 = fib_m
            fib_m = fib_m2 + fib_m1
        end do

        offset = 0
        index = -1

        do while (fib_m > 1)
            i = min(offset + fib_m2, n)

            if (arr(i) < target) then
                fib_m = fib_m1
                fib_m1 = fib_m2
                fib_m2 = fib_m - fib_m1
                offset = i
            else if (arr(i) > target) then
                fib_m = fib_m2
                fib_m1 = fib_m1 - fib_m2
                fib_m2 = fib_m - fib_m1
            else
                index = i
                return
            end if
        end do

        if (fib_m1 == 1 .and. offset + 1 <= n .and. arr(offset + 1) == target) then
            index = offset + 1
        end if
    end function fibonacci_search

    ! ===================================================================
    ! PERFORMANCE COMPARISON
    ! ===================================================================

    subroutine performance_comparison(arr, n)
        integer, intent(in) :: arr(:), n
        integer :: target, i, result
        real(8) :: start, finish
        real(8) :: time_binary, time_interpolation, time_jump, time_exponential
        integer, parameter :: num_searches = 10000

        print '(A)', 'Performance Comparison'
        print '(A)', repeat('=', 75)
        print '(A, I0, A)', 'Performing ', num_searches, ' searches...'
        print *

        target = n  ! Search for last element (worst case for many algorithms)

        ! Binary Search
        call cpu_time(start)
        do i = 1, num_searches
            result = binary_search_iterative(arr, n, target)
        end do
        call cpu_time(finish)
        time_binary = finish - start

        ! Interpolation Search
        call cpu_time(start)
        do i = 1, num_searches
            result = interpolation_search(arr, n, target)
        end do
        call cpu_time(finish)
        time_interpolation = finish - start

        ! Jump Search
        call cpu_time(start)
        do i = 1, num_searches
            result = jump_search(arr, n, target)
        end do
        call cpu_time(finish)
        time_jump = finish - start

        ! Exponential Search
        call cpu_time(start)
        do i = 1, num_searches
            result = exponential_search(arr, n, target)
        end do
        call cpu_time(finish)
        time_exponential = finish - start

        ! Print results
        print '(A)', 'Algorithm              Time (s)    Relative Speed'
        print '(A)', repeat('-', 55)
        print '(A, F10.6, A)', 'Binary Search          ', time_binary, '    1.00x (baseline)'
        print '(A, F10.6, A, F6.2, A)', 'Interpolation Search   ', time_interpolation, &
              '    ', time_binary / time_interpolation, 'x'
        print '(A, F10.6, A, F6.2, A)', 'Jump Search            ', time_jump, &
              '    ', time_binary / time_jump, 'x'
        print '(A, F10.6, A, F6.2, A)', 'Exponential Search     ', time_exponential, &
              '    ', time_binary / time_exponential, 'x'
        print *
    end subroutine performance_comparison

    ! ===================================================================
    ! UTILITY FUNCTIONS
    ! ===================================================================

    subroutine print_result(index, target, algorithm)
        integer, intent(in) :: index, target
        character(len=*), intent(in) :: algorithm

        if (index /= -1) then
            print '(A, A, I0, A, I0)', trim(algorithm), ' search: Found ', &
                  target, ' at index ', index
        else
            print '(A, A, I0, A)', trim(algorithm), ' search: ', target, ' not found'
        end if
    end subroutine print_result

end program advanced_search
